src/poligrain/spatial.py

```diff
@@ -477,3 +478,128 @@ def get_grid_time_series_at_intersections(grid_data, intersect_weights):
+
+
+def get_closest_points_to_line(ds_cmls, ds_gauges, max_distance, n_closest):
+    """Get closest points to line.
+
+    Finds n closest points from a CML within given max distance. Note that the
+    function guarantees that all returned points are within max distance to
+    the CML, not that all points that are within max distance are returned. Uses
+    KDTree for fast processing of large datasets.
+
+    Parameters
+    ----------
+    ds_cmls: xarray.Dataset
+        Dataset of line data using the OpenSense naming convention for CMLs. It
+        must contain the coordinate cml_id with the cml names. It must also
+        contain projected coordinates site_0_y, site_0_x, site_1_y and site_1_x
+        as well as the CML length.
+    ds_gauges: xarray.Dataset
+        Dataset of point data using the OpenSense data format conventions for PWS.
+        The dataset must contain the coordinate 'id' with the PWS names. It must
+        also contain projected coordinates x and y.
+    max_distance: float
+        Maximum distance a point can have to the CML, measured as the smallest
+        distance from the point to the line. Points outside this range is not
+        considered close to the CML.
+    n_closest: int
+        Maximum number of points that are returned.
+
+    Returns
+    -------
+    closest_gauges: xarray.Dataset
+        Dataset with CML ids and corresponding n_closest point names and distance.
+        If a CML has less that "n_closest" nearby points, the reminding entries
+        are filled with 'nan'.
+
+    """
+    # transfer raingauge and CML coordinates to numpy, for faster access in loop
+    coords_cml_a = np.hstack(
+        [ds_cmls.site_0_y.data.reshape(-1, 1), ds_cmls.site_0_x.data.reshape(-1, 1)]
+    )
+
+    coords_cml_b = np.hstack(
+        [ds_cmls.site_1_y.data.reshape(-1, 1), ds_cmls.site_1_x.data.reshape(-1, 1)]
+    )
+
+    coords_gauge = np.hstack(
+        [ds_gauges.y.data.reshape(-1, 1), ds_gauges.x.data.reshape(-1, 1)]
+    )
+
+    cml_lengths = ds_cmls.length.data / 2
+
+    # calculate CML midpoints by using the average of site a and b
+    coords_cml = np.hstack(
+        [
+            ((coords_cml_a[:, 0] + coords_cml_b[:, 0]) / 2).reshape(-1, 1),
+            ((coords_cml_a[:, 1] + coords_cml_b[:, 1]) / 2).reshape(-1, 1),
+        ]
+    )
+
+    # extend max_distance to include diagonal length
+    max_distance_add = np.sqrt(2 * max_distance**2)
+
+    # array for storing indices of gauges close to CML, -1 represent no gauge
+    list_gauges = np.ones([ds_cmls.cml_id.size, n_closest]).astype(int) * (-1)
+
+    # array for storing distances from CML to close gauges
+    cml_gauge_dist = np.zeros([ds_cmls.cml_id.size, n_closest]) * np.nan
+
+    # create KDTree object for all gauges
+    kd_tree2 = KDTree(coords_gauge)
+
+    for i in range(len(ds_cmls.cml_id)):
+        # create KDTree object for midpoint of CML_i
+        kd_tree1 = KDTree([coords_cml[i]])  # select cml i
```

Comment:
I think it would be easier to just use `kd_tree2.query(coords_cml[i], max_distance, etc...)` instead of creating the second tree `kd_tree1` and then do a sparse distance matrix calculations.

Reply:
...it is actually not super important how you do it, as long as there is not significant issue with computation time for larger datasets.

Reply:
I played around with kd_tree2.query but it requires that we specify the number of gauges to return. Since we at a later stage do a more precise calculation of the distances (using shapely)  kd_tree2.query might not return the same closest gauge as we would get using shapely. Instead I opted to use kd_tree.query_ball_point which should behave similar but finds all gauges within max distance.

Reply:
both kd_tree2.query and kd_tree.query_ball_point runs faster and makes the code more readable compared to using kd_tree1.sparse_distance_matrix